pallets/rewards/src/lib.rs

```diff
@@ -58,148 +58,134 @@ mod mock;
 
 #[frame_support::pallet]
 pub mod pallet {
 	use super::*;
 
 	#[pallet::config]
-	pub trait Config: frame_system::Config {
-		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Config>::Event>;
 
 		/// Identifier of this pallet used as an account where stores the reward that is not claimed.
 		/// When you distribute reward, the amount distributed goes here.
 		#[pallet::constant]
 		type PalletId: Get<PalletId>;
 
 		/// Type used to identify domains.
-		type DomainId: TypeInfo
-			+ MaxEncodedLen
-			+ codec::FullCodec
-			+ Copy
-			+ PartialEq
-			+ sp_std::fmt::Debug;
+		type DomainId: TypeInfo + MaxEncodedLen + FullCodec + Copy + PartialEq + sp_std::fmt::Debug;
 
 		/// Type used to identify currencies.
 		type CurrencyId: AssetId + MaxEncodedLen;
 
 		/// Identifier for the currency used to give the reward.
 		type RewardCurrency: Get<Self::CurrencyId>;
 
-		/// Type used to handle balances.
-		type Balance: Balance + MaxEncodedLen + FixedPointOperand + TryFrom<Self::SignedBalance>;
+		/// Type used to identify groups.
+		type GroupId: FullCodec + TypeInfo + MaxEncodedLen + Copy + PartialEq + sp_std::fmt::Debug;
 
 		/// Type used to handle currency transfers and reservations.
-		type Currency: MutateHold<Self::AccountId, AssetId = Self::CurrencyId, Balance = Self::Balance>
-			+ Mutate<Self::AccountId, AssetId = Self::CurrencyId, Balance = Self::Balance>;
-
-		/// Type used to handle a Balance that can have negative values
-		type SignedBalance: TryFrom<Self::Balance>
-			+ codec::FullCodec
-			+ Copy
-			+ Default
-			+ TypeInfo
-			+ MaxEncodedLen
-			+ Signed
-			+ FixedPointOperand
-			+ EnsureAdd
-			+ EnsureSub;
-
-		/// Type used to handle rates as fixed points numbers.
-		type Rate: FixedPointNumber + TypeInfo + MaxEncodedLen + Encode + Decode;
+		type Currency: MutateHold<Self::AccountId, AssetId = Self::CurrencyId, Balance = BalanceOf<Self, I>>
+			+ Mutate<Self::AccountId, AssetId = Self::CurrencyId, Balance = BalanceOf<Self, I>>;
 
-		/// Type used to identify groups.
-		type GroupId: codec::FullCodec
-			+ TypeInfo
-			+ MaxEncodedLen
-			+ Copy
-			+ PartialEq
-			+ sp_std::fmt::Debug;
-
-		/// Max number of currency movements. See [`Rewards::attach_currency()`].
-		#[pallet::constant]
-		type MaxCurrencyMovements: Get<u32> + TypeInfo;
+		/// Specify the internal reward mechanism used by this pallet.
+		/// Check available mechanisms at [`mechanism`] module.
+		type RewardMechanism: RewardMechanism;
 	}
 
 	#[pallet::pallet]
 	#[pallet::generate_store(pub(super) trait Store)]
-	pub struct Pallet<T>(_);
+	pub struct Pallet<T, I = ()>(_);
 
 	// --------------------------
 	//          Storage
 	// --------------------------
 
 	#[pallet::storage]
-	pub(super) type Currencies<T: Config> = StorageMap<
+	pub(super) type CurrencyGroup<T: Config<I>, I: 'static = ()> =
```

Comment:
Initially, it was joined, but I needed to split it because, from the pallet perspective, you do not know what a currency is. Also, the group id of a currency does not depend on how you implement the currency. It is simply some id used by the pallet (not the mechanism) to link currencies to groups. A mechanism without currency movement still needs to have a group id attached to its currencies, you could not move currencies but you still need to associate the currency to a group the first time.

Reply:
Let me rethink this because I think maybe it is possible to put both under the same map as a tuple or similar.

Reply:
Done!